### PR TITLE
Fix OpenSSL error handling for EOF (support for OpenSSL 3.2)

### PR DESCRIPTION
**Pull request by straight-shoota**
This is following the reasoning from https://github.com/crystal-lang/crystal/issues/14168#issuecomment-1877224928

I'm not entirely sure this is exactly correct API usage. The documentation is not very extensive. But judging from the resulting behaviour, and backed by available documentation it seems to be more correct than the current implementation.

### CODE REVIEW

**Comment by HertzDevil** [comment]
On Windows it looks like you have to check `WinError.wsa_value` instead

File: src/openssl.cr [hunk 1]
```diff
@@ -89,15 +89,13 @@ module OpenSSL
         when .syscall?
           @code, message = fetch_error_details
           if @code == 0
-            case return_code
-            when 0
+            errno = {% if flag?(:win32?) %} WinError.wsa_value {% else %} Errno.value {% end %}
+            if {% if flag?(:win32) %} errno.error_success? {% else %} errno.none? {% end %}
```
Comment (beta-ziliani) [inline src/openssl.cr:93]: Really a nitpick, but that double `if {% if` is hard to parse
```suggestion
            success = {% if flag?(:win32) %} errno.error_success? {% else %} errno.none? {% end %}
            if success
```

**Comment by beta-ziliani** [comment]
Can't we add OpenSSL 3.2 to the OpenSSL CI to check this change?

**Comment by straight-shoota** [comment]
We could, but it requires some non-trivial additions to the CI configuration.
Our tests are all running in the docker image of the previous release. At the moment that's based on alpine 3.17, for which OpenSSL 3.2 is not available. Nor is OpenSSL 3.1, so we're not explicitly testing that version either.

**Review by beta-ziliani — APPROVED** [review]

**Comment by straight-shoota** [comment]
This got eventually reverted in #14217 due to https://github.com/crystal-lang/crystal/issues/14200#issuecomment-1887112353
